Source/NSubstitute/Core/PendingSpecification.cs

```diff
@@ -1,29 +1,65 @@
+using System;
+
 namespace NSubstitute.Core
 {
     public class PendingSpecification : IPendingSpecification
     {
-        ICallSpecification _pendingSpec;
+        private readonly ISubstitutionContext _substitutionContext;
+
+        public PendingSpecification(ISubstitutionContext substitutionContext)
+        {
+            _substitutionContext = substitutionContext;
+        }
+
+        public bool HasPendingCallSpecInfo()
+        {
+            //Check ownership to avoid potential issues when spec from foreign substitution is present.
+            var currentInfo = _substitutionContext.ThreadLocalPendingSpecificationInfo as PendingSpecificationInfo;
+            return currentInfo != null && currentInfo.Owner == this;
+        }
 
-        public bool HasPendingCallSpec()
+        public IPendingSpecificationInfo UseCallSpecInfo()
         {
-            return _pendingSpec != null;
+            var info = _substitutionContext.ThreadLocalPendingSpecificationInfo as PendingSpecificationInfo;
+
+            if (info != null && info.Owner != this)
+            {
+                //That should never happen. It could only happen if there are some bugs or unexpected usages.
+                throw new InvalidOperationException("Specification from foreign substitution is present on stack.");
+            }
+
+            Clear();
+
+            return info;
         }
 
-        public ICallSpecification UseCallSpec()
+        public void SetCallSpecification(ICallSpecification callSpecification)
         {
-            var specToUse = _pendingSpec;
-            _pendingSpec = null;
-            return specToUse;
+            _substitutionContext.ThreadLocalPendingSpecificationInfo = new PendingSpecificationInfo(callSpecification, null, this);
+        }
+
+        public void SetLastCall(ICall lastCall)
+        {
+            _substitutionContext.ThreadLocalPendingSpecificationInfo = new PendingSpecificationInfo(null, lastCall, this);
         }
 
         public void Clear()
         {
-            UseCallSpec();
+            _substitutionContext.ThreadLocalPendingSpecificationInfo = null;
         }
 
-        public void Set(ICallSpecification callSpecification)
+        private class PendingSpecificationInfo: IPendingSpecificationInfo
         {
-            _pendingSpec = callSpecification;
+            public ICallSpecification CallSpecification { get; }
+            public ICall LastCall { get; }
+            public IPendingSpecification Owner { get; }
+
+            public PendingSpecificationInfo(ICallSpecification callSpecification, ICall lastCall, IPendingSpecification owner)
```

Comment:
I tend to keep things simple. Thread local wrapper already improves the situation a lot. I think it will be sufficient for now. I don't remember any related bug report and cannot imagine a situation when it can happen. I know everything is possible but I'm for simplicity here :)